WaitForPodsReady.recoveryTimeout implementation

PBundyra opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
It introduces a new logic to WaitForPodsReady feature, to support new `.recoveryTimeout` API, according to [this KEP](https://github.com/kubernetes-sigs/kueue/pull/2737)

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #2732

#### Special notes for your reviewer:
This is one of two PRs for this feature. The [first one](https://github.com/kubernetes-sigs/kueue/pull/4302) introduces API only, and this one adds logic.
Please wait until the PR with API is merged

The integration test in this PR should be enough for the Kueue part, but I'd like to have an additional e2e test in the follow up, which covers full journey with the batchv1.Job controller 

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Kueue exposes a new recovery mechanism as part of the WaitForPodsReady API. This evicts jobs which surpasses configured threshold for pod's recovery during runtime
```

PBundyra commented:
@mimowo @tenzen-y 
Can you PTAL? The API PR has been merged, I rebased this one and it's ready for review

tenzen-y commented:
> @mimowo @tenzen-y Can you PTAL? The API PR has been merged, I rebased this one and it's ready for review

ACK
/assign

mimowo commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -460,15 +460,17 @@ func (r *JobReconciler) ReconcileGenericJob(ctx context.Context, req ctrl.Reques
-		if !apimeta.IsStatusConditionPresentAndEqual(wl.Status.Conditions, condition.Type, condition.Status) {
-			log.V(3).Info(fmt.Sprintf("Updating the PodsReady condition with status: %v", condition.Status))
+		condition := generatePodsReadyCondition(log, job, wl)
+		if !workload.HasCondition(wl, &condition) {
```
I think the semantics of the HasCondition is quite unclear just by looking at its name. In particular, should we check timestamps, message, reason etc. are the obvious questions. 

I think it would be better to have HasConditionWithTypeAndReason (I believe we should already have something like this). 

I did a quick check and the HasCondition function checks the message, I think we should avoid building any logic on the messages. They are only for humans.

mimowo replied:
> There are two cases in which the message is the only differentiator between conditions:

Thanks for explaining, I see the motivation.

> Are you suggesting merging those two and setting more generic message?

Possibly, but not necessarily. I can see at least two options:
1. make the message generic IIUC this is how it was done before 
2. discriminate against the two states by the reason field, e.g. `WaitForPodsStart` and `NotAdmitted`

Basically, we should not build any logic based on messages being equal or different, because messages are free to change at any point. We have the analogous policy in the code k8s.

My preference is for (2.), but since it was done like in (1.) I'm ok to leave it for follow up.

PBundyra replied:
I also like the second approach better. I've added `NotAdmitted` reason and removed message comparison from the `HasConditionWithTypeAndReason` function

mimowo commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -1148,22 +1150,72 @@ func (r *JobReconciler) ignoreUnretryableError(log logr.Logger, err error) error
+
+	podsReadyStartCond := workload.IsPodsReadyStartCond(wl)
+	podsReadyRecoveryCond := workload.IsPodsReadyRecoveryCond(wl)
+	log.V(2).Info("PATRYK before switch",
```
Revisit the log. Surely remove PATRYK, but also probably this is V(3). I also think the ":" aren't necessary. Isn't the log folded as json?

mimowo commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -1148,22 +1150,72 @@ func (r *JobReconciler) ignoreUnretryableError(log logr.Logger, err error) error
+		status = metav1.ConditionTrue
+		message = "All pods reached readiness and the job is running"
+	default:
+		log.V(2).Info("Unexpected PodsReady condition",
```
Here probably v(2) is justified, but clear the ":". Also, maybe add "WARNING" if this is unexpected.

PBundyra replied:
Done

mimowo reviewed (commented):
ack, initiall pass, I will continue next week

mimowo commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -1148,22 +1150,72 @@ func (r *JobReconciler) ignoreUnretryableError(log logr.Logger, err error) error
+	switch {
+	// 1a. The workload has just been admitted and doesn't have PodsReady cond yet
+	// 1b. The workload has just been admitted, and not all pods are ready
+	case !job.PodsReady() && (apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadPodsReady) == nil ||
```
Make sure PodsReady is computed only once in this function. While it is fast for batch/Job it is slower for JobSet (where it depends on the number of PodSets), but in general it depends on the integration code, so could be non-constant.

PBundyra replied:
Good point, done

mimowo commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -1148,22 +1150,72 @@ func (r *JobReconciler) ignoreUnretryableError(log logr.Logger, err error) error
+		podsReadyRecoveryCond != nil):
+		reason = kueue.WorkloadWaitForPodsRecovery
+		status = metav1.ConditionFalse
+		message = "One of pods has failed and is waiting for recovery"
```
```suggestion
		message = "At least one pod has failed, waiting for recovery"
```
proposal, because the previous message suggested it is the pod that is waiting for recovery, whilst the pod will never recover :). What can recover is the workload.  We could maybe say "At least one pod has failed, workload waiting for recovery", but up to you. I'm also open to other proposals.

PBundyra replied:
I agree it was misleading, done

mimowo commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -1148,22 +1150,72 @@ func (r *JobReconciler) ignoreUnretryableError(log logr.Logger, err error) error
+	case job.PodsReady() && podsReadyStartCond != nil:
+		reason = kueue.WorkloadWaitForPodsStart
+		status = metav1.ConditionTrue
+		message = "All pods reached readiness and the job is running"
```
Maybe it is safer not to say "job", but rather workload? The reason is that Kueue already supports serving workloads such as Deployments, StatefulSet or LWS. For them "Job" might be misleading. I'm aware we use "job" in many places like that for historical reasons, but I want to migrate from it over time. PTAL https://github.com/kubernetes-sigs/kueue/issues/4286

PBundyra replied:
Thanks, done

mimowo commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -1148,22 +1150,72 @@ func (r *JobReconciler) ignoreUnretryableError(log logr.Logger, err error) error
+	if !workload.IsAdmitted(wl) {
+		// The workload has not been admitted yet
+		// or it was admitted in the past but it's evicted/requeued
+		return metav1.Condition{
+			Type:    kueue.WorkloadPodsReady,
+			Status:  status,
+			Reason:  reason,
+			Message: message,
+			// ObservedGeneration is added via workload.UpdateStatus
+		}
```
This is duplicated with the last return. Can we maybe create a function such as `createPodsReadyCondition` and call it from here, but also from the switch cases, instead of the final return.

PBundyra replied:
PTAL if it aligns with your suggestion

mimowo commented on pkg/controller/jobs/job/job_controller_test.go:
```diff
@@ -491,6 +494,143 @@ func TestReconciler(t *testing.T) {
 		wantEvents        []utiltesting.EventRecord
 		wantErr           error
 	}{
+		"PodsReady is set to False before Workload is Admitted": {
```
Do the test cases cover 100% of the code in `generatePodsReadyCondition`? You can generate the coverage data by adding `-coverprofile=` while running unit tests, then report with `go tool cover -html`.

PBundyra replied:
It covers 95.8% of the code, the remaining ~4% is because we don't test the `default` code of block when the conditions are invalid

mimowo replied:
Thanks for checking. 

> we don't test the default code of block when the conditions are invalid

What does it mean the conditions are invalid? Does it mean the "reason" field is unrecognized? Is this the scenario which will happen to every workload on upgrade? If so then I believe it is worth covering. 

Actually, I believe we should have some form of a test demonstrating the behavior on upgrade. We have two options: keep the old reason, but make sure it is handled as WaitForPodsStart, or convert it to "WaitForPodsStart" on the first iteration. I would be leaning to convert it actually.

PBundyra replied:
We do convert it on the first iteration, I'll cover that in tests

PBundyra replied:
Added the test, covering 100% now

mimowo reviewed (commented):
Completed the first pass. I will revisit when the remarks are addressed.

PBundyra commented on pkg/controller/core/workload_controller.go:
```diff
@@ -826,8 +826,21 @@ func (r *WorkloadReconciler) admittedNotReadyWorkload(wl *kueue.Workload) (bool,
-	admittedCond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadAdmitted)
-	elapsedTime := r.clock.Since(admittedCond.LastTransitionTime.Time)
+
+	var elapsedTime time.Duration
```
PTAL if this is what you had in mind

mimowo replied:
Pretty much, and not you don't need to declare elapsed time outside of the switch, right?

PBundyra replied:
Right, I've cleaned this up

mimowo commented on pkg/workload/workload.go:
```diff
@@ -771,6 +771,22 @@ func IsActive(w *kueue.Workload) bool {
 	return ptr.Deref(w.Spec.Active, true)
 }
 
+func IsPodsReadyStartCond(w *kueue.Workload) *metav1.Condition {
```
A function with a name starting with `Is` should return bool, so something is off here. 

Maybe we can introduce a helper function `FindConditionWithReason`? It could already replace places like this in a follow up: https://github.com/kubernetes-sigs/kueue/blob/a70fd62ad7649877147a150afa92725e366a01f5/pkg/workload/workload.go#L725-L727

PBundyra replied:
Added the function

mimowo commented on pkg/workload/workload.go:
```diff
@@ -771,6 +771,22 @@ func IsActive(w *kueue.Workload) bool {
 }
 
+func IsPodsReadyStartCond(w *kueue.Workload) *metav1.Condition {
+	cond := apimeta.FindStatusCondition(w.Status.Conditions, kueue.WorkloadPodsReady)
+	if cond == nil || cond.Reason != kueue.WorkloadWaitForPodsStart {
+		return nil
+	}
+	return cond
+}
```
```suggestion
	if cond := apimeta.FindStatusCondition(w.Status.Conditions, kueue.WorkloadPodsReady); 
	   cond != nil && cond.Reason == kueue.WorkloadWaitForPodsStart {
		return cond
	}
	return nil
}
```
nit, to avoid negating the reason equality which makes it harder to imo.

PBundyra replied:
Done

mimowo commented on apis/kueue/v1beta1/workload_types.go:
```diff
@@ -587,6 +587,9 @@ const (
 	// and workload waits for recovering.
 	WorkloadWaitForPodsRecovery = "WaitForPodsRecovery"
+
+	// WorkloadNotAdmitted is a reason that indicates the Workload has not been admitted
```
Is this reason used for workload which get evicted and are waiting to be re-admitted?
If no, then maybe say "the Workload has never been admitted yet" to make it more explicit, wdyt?

PBundyra replied:
It is used for both cases - first admission and readmission

mimowo replied:
So, then maybe say "... the Workload is not admitted". I'm not an English expert, but " has not been admitted" sounds ambiguous to me, and I actually assumed you mean "has never been admitted yet".

PBundyra replied:
Oh yes, done

tenzen-y commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -1148,24 +1150,62 @@ func (r *JobReconciler) ignoreUnretryableError(log logr.Logger, err error) error
-		Message: message,
-		// ObservedGeneration is added via workload.UpdateStatus
+func generatePodsReadyCondition(log logr.Logger, job GenericJob, wl *kueue.Workload) metav1.Condition {
+	if !workload.IsAdmitted(wl) {
+		// The workload has not been admitted yet
+		// or it was admitted in the past but it's evicted/requeued
+		return workload.CreatePodsReadyCondition(metav1.ConditionFalse,
+			kueue.WorkloadNotAdmitted,
+			"Workload isn't admitted")
+	}
```
I do not think that this is needed since we will do as following, IIUC:

- `it was admitted in the past but it's evicted/requeued`: kueue turns the QuotaReserved condition to False and removes Admitted condition from Workload.
- `The workload has not been admitted yet`: We tell by the `Admitted` and `QuotaReserved` conditions whether or not the Workload is admitted. What is the value of an additional admitted state rather than `Admitted` and `QuotaReserved`.

PBundyra replied:
My primary motivation was that after creating a Workload we set the PodsReady condition to False, even tho it hasn't been admitted yet, and is not counting down. I thought it can be misleading so I wanted to differentiate the case when the workload is not admitted yet, and when it's admitted and counting down

tenzen-y replied:
In that case could we clear the `PodsReady` condition if the Workload has it since `{Admitted: nil, QuotaReserved: false, PodsReady: false}` is weird.

The `{QuotaReserved: False}` covers all cases for `WorkloadNotAdmitted`. I mean we represent non admitted workloads by `{QuotaRerserved: False}` and `WorkloadNonAdmitted` reason. I think we should represent the state by only one semantics.

Do you have any logic to verify something by this `WorkloadNonAdmitted` reason?

PBundyra replied:
> In that case could we clear the `PodsReady` condition if the Workload has it since `{Admitted: nil, QuotaReserved: false, PodsReady: false}` is weird.

Just so you know, this is the current Kueue's behavior. 

Could you elaborate on what you mean by "clearing" the condition? Do you mean to set empty strings in Condition's Status, Reason and Message?

PBundyra replied:
> After this PR, `Admitted: True` or `PodsReady: False, Reason: NonAdmitted` can represent admission status.

I see, that's a good catch.

> In conclusion, basically, I think we should represent Admission state only by Admitted and QuotaReserved.

@mimowo are you fine with getting rid of the `NotAdmitted` reason and sticking with the 1st approach discussed [here](https://github.com/kubernetes-sigs/kueue/pull/4301#discussion_r1967304906)?

mimowo replied:
The NotAdmitted reason is not a new way to represent not admitted workload. This remains solely dependent on the Admitted condition type.

The reason is only used within the context of the PodsReady condition to indicate why PodsReady is False.

That being said I'm ok to not introduce the reason as part of this PR because it is a bit orthogonal and it wasn't there ok. I still think it makes sense as a follow up

PBundyra replied:
SGTM, I'll revert the changes related to NotAdmitted, and will create a new issue once this PR is merged

tenzen-y commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -1148,24 +1150,62 @@ func (r *JobReconciler) ignoreUnretryableError(log logr.Logger, err error) error
+			kueue.WorkloadWaitForPodsRecovery,
+			"All pods reached readiness and the workload is running")
+	default:
+		log.V(2).Info("WARNING Unexpected PodsReady condition",
```
What is the assumed situations where we face here in this flow?
https://github.com/PBundyra/kueue/blob/daad59ea89bc520b9af7d3ed58c4117bf03ec337/keps/349-all-or-nothing/README.md

tenzen-y replied:
Sorry for your confusion. This is just a question to clarify when we face this path (`default`).

PBundyra replied:
Now the `default` path is only for unexpected circumstances. Also, we need to return something if non of the above `cases` are true

PBundyra commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -1148,24 +1150,62 @@ func (r *JobReconciler) ignoreUnretryableError(log logr.Logger, err error) error
+			"Workload isn't admitted")
+	}
+
+	podsReadyStartCond := workload.FindStatusConditionWithReason(wl, kueue.WorkloadPodsReady, kueue.WorkloadWaitForPodsStart)
+	podsReadyRecoveryCond := workload.FindStatusConditionWithReason(wl, kueue.WorkloadPodsReady, kueue.WorkloadWaitForPodsRecovery)
```
Do you mean something along the lines:

```golang
var podsReadyStartCond, podsReadyRecoveryCond *metav1.Condition
podsReadyCond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadPodsReady)
if podsReadyCond != nil && podsReadyCond.Reason == kueue.WorkloadWaitForPodsStart {
	podsReadyStartCond = podsReadyCond
} else if podsReadyCond != nil && podsReadyCond.Reason == kueue.WorkloadWaitForPodsRecovery {
	podsReadyRecoveryCond = podsReadyCond
}
```

mimowo replied:
no, I mean not to have the `podsReadyStartCond, podsReadyRecoveryCond ` variables at all.

mimowo replied:
In order words to have something like:

```golang
podsReadyCond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadPodsReady)
if podsReadyCond == nil {
 // handle case without the condition
}
reason := podsReadyCond.Reason
status := podsReadyCond.Status
switch {
  case reason == XYZ && status == ABC {
     ...
  }
  ...
```

mimowo replied:
Yeah, almost, this looks better. 

However, I still have a hard time to see if the switch cases are complementary or not, as each of them has a couple of entry checks. 

Maybe it could be helpful to at least check podsReadyCond == nil as a preliminary check so that it does not need to be checked in every case down the line?

PBundyra replied:
I see, ptal

tenzen-y commented on pkg/controller/core/workload_controller.go:
```diff
@@ -826,9 +826,20 @@ func (r *WorkloadReconciler) admittedNotReadyWorkload(wl *kueue.Workload) (bool,
+		admittedCond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadAdmitted)
+		elapsedTime := r.clock.Since(admittedCond.LastTransitionTime.Time)
+		return true, max(r.waitForPodsReady.timeout-elapsedTime, 0)
+	} else {
```
```suggestion
	}
```

This `else` is redundant.

tenzen-y commented on pkg/controller/core/workload_controller.go:
```diff
@@ -826,9 +826,20 @@ func (r *WorkloadReconciler) admittedNotReadyWorkload(wl *kueue.Workload) (bool,
+		elapsedTime := r.clock.Since(admittedCond.LastTransitionTime.Time)
+		return true, max(r.waitForPodsReady.timeout-elapsedTime, 0)
+	} else {
+		// podsReadyCond.Reason == kueue.WorkloadWaitForPodsRecovery
+		// A pod has failed and the workload is waiting for recovery
```
What if `podReadyCond` has `reason: PodsReady` and `status: False`?

PBundyra replied:
With the current flow `podsReadyCond` with `PodsReady` reason will be converted into `WaitForPodsStart` here:
https://github.com/kubernetes-sigs/kueue/pull/4301/files#diff-552ab657667de70900b502e2617bcb405fbf4203ff08b2a35d7c87f2d637a75bR1196-R1202

tenzen-y commented on pkg/controller/core/workload_controller.go:
```diff
@@ -826,9 +826,20 @@ func (r *WorkloadReconciler) admittedNotReadyWorkload(wl *kueue.Workload) (bool,
-	elapsedTime := r.clock.Since(admittedCond.LastTransitionTime.Time)
-	return true, max(r.waitForPodsReady.timeout-elapsedTime, 0)
+
+	if podsReadyCond == nil || podsReadyCond.Reason == kueue.WorkloadWaitForPodsStart {
```
```suggestion
	if podsReadyCond == nil || podsReadyCond.Reason == kueue.WorkloadWaitForPodsStart || podsReadyCond.Reason == "PodsReady" {
```

IIUC, the `WaitForPodsStart` is compatible with the old `PodsReady` reason.

> We introduce new WorkloadWaitForPodsStart and WorkloadWaitForPodsRecovery reasons to distinguish the reasons of setting the PodsReady=false condition.

PBundyra replied:
We don't necessarily need to cover that case as we convert the old reason into a new one in the `generatePodsReadyCondition` function. However if you think it would be safer to take into account the old reason in both places I'm happy to update that

tenzen-y replied:
It would be better to confirm both places (JobFramework and WorkloadController (here)) since they usually deploy kueue-external-manager to support non-upstream Jobs. In that case, the kueue-external-manager might be older than kueue-manger, which means kueue-external-manager JobFramework reconciler keeps repling on PodsReady reason.

It might be better to do following steps:

v0.11 (upcoming latest release): Check if podsReady reason is PodsReady in the workload-controller as well, which means verifying PodsReady with PodsReady reason case both in JobFramwork and WrokloadController. And add `[ACTION REQUIRED]` release note to upgrade JobFramwrork version in kueue-external-manager.
v0.12: Remove the redundant verification from WorkloadController, and then keep it only in JobFramework.

@mimowo Any thoughts?

tenzen-y replied:
> This problem is more visible in case of external controllers, but is also present within build-in Kueue controllers, which is also why I'm thinking some integration test would be great.

SGTM, in that case, let's do 2 things in this PR:

1. Keep previous cases for PodsReady with `PodsReady` reason
2. Add Integration testings.

tenzen-y commented on pkg/workload/workload.go:
```diff
@@ -798,6 +807,28 @@ func IsEvicted(w *kueue.Workload) bool {
+// HasConditionWithTypeAndReason checks if there is a condition in Workload's status
+// with exactly the same Type, Status and Reason
+func HasConditionWithTypeAndReason(w *kueue.Workload, cond *metav1.Condition) bool {
+	for _, statusCond := range w.Status.Conditions {
+		if statusCond.Type == cond.Type && statusCond.Reason == cond.Reason &&
+			statusCond.Status == cond.Status {
+			return true
+		}
+	}
+	return false
```
```suggestion
	condition := FindStatusConditionWithReason(w, cond.Type, cond.reason)
	return condition != nil && condition.Status == cond.Status
```

I think this could leverage `FindStatusConditionWithReason`.

tenzen-y reviewed (commented):
Could we leave test cases to verify compatibility for older `PodsReady` reason?

> We introduce new WorkloadWaitForPodsStart and WorkloadWaitForPodsRecovery reasons to distinguish the reasons of setting the PodsReady=false condition.

https://github.com/PBundyra/kueue/blob/daad59ea89bc520b9af7d3ed58c4117bf03ec337/keps/349-all-or-nothing/README.md#timeout-on-reaching-the-podsready-condition

mimowo commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -1148,24 +1150,56 @@ func (r *JobReconciler) ignoreUnretryableError(log logr.Logger, err error) error
+	switch {
+	// 1a. The workload has just been admitted and doesn't have PodsReady cond yet
+	// 1b. The workload has just been admitted, and not all pods are ready
+	case !podsReady && (podsReadyCond == nil || podsReadyCond.Reason == kueue.WorkloadWaitForPodsStart && podsReadyCond.Status == metav1.ConditionFalse):
```
Actually, is this even correct, what happens if `podsReady=true`, but podsReadyCond == nil. IIUC the code will enter the default, which will set the condition to False. This does not seem accurate since `podsReady=True`.

PBundyra replied:
Fixed, added a test scenario covering that

PBundyra commented:
> Could we leave test cases to verify compatibility for older `PodsReady` reason?
> 
> > We introduce new WorkloadWaitForPodsStart and WorkloadWaitForPodsRecovery reasons to distinguish the reasons of setting the PodsReady=false condition.
> 
> https://github.com/PBundyra/kueue/blob/daad59ea89bc520b9af7d3ed58c4117bf03ec337/keps/349-all-or-nothing/README.md#timeout-on-reaching-the-podsready-condition

Do you mean something different than this one?

https://github.com/kubernetes-sigs/kueue/pull/4301/files#diff-4691f767e133e2011fb5255af68af22a19112c49cad441049aa4a0ef65ba5ebdR634-R660

PBundyra commented:
/retest

mimowo commented:
LGTM overall, what I'm missing yet is an integration test for the baseline scenario.

I think it would be valuable to ensure the mechanism works well spanning multiple reconciles. I'm ok to leave it for a follow up, especially if you tested this locally on a running cluster.

tenzen-y commented on test/integration/singlecluster/scheduler/podsready/scheduler_test.go:
I think that we want to add integration test cases, at least.

mimowo commented:
/hold
for the https://github.com/kubernetes-sigs/kueue/pull/4301#discussion_r1967483299 comment which I consider blocking.
The integration tests are block-ish, but to me could be added in a quick follow up if it was tested manually end-to-end on a local cluster.

PBundyra commented:
> I meant could we keep holding cases where input Workload has PodsReady condition with PodsReady reason? For example, this PR replaces those cases like [this](https://github.com/kubernetes-sigs/kueue/pull/4301/files#diff-fbea57b850eb71f650b095afe623b9ed3e9ce3eb5cc35ab965e3498a010a17e0R1114). But, It would be better to keep holding original case as long as the behavior is supported.

Would you like to leave all the old test cases and duplicate all of them with the only difference being the new reason? Or maybe just leave them as they were without duplicating?

PBundyra commented:
@mimowo @tenzen-y 
Here's an integration test that checks the baseline scenario with a failing pod: https://github.com/kubernetes-sigs/kueue/pull/4301/files#diff-f1db0eb3182b0e49163aaa7fc1fcc2ee21f01eda279797cc07d01fd16da41ce6R2182-R2239

Would you like to have another one that tests anything particular that is not covered by this one?
I've chosen to put the test into batchv1.Job tests as it's easy to controller the `PodsReady()` method from the framework interface

PBundyra commented on pkg/controller/core/workload_controller_test.go:
```diff
@@ -1081,7 +1151,7 @@ func TestReconcile(t *testing.T) {
 				Condition(metav1.Condition{
 					Type:    kueue.WorkloadPodsReady,
 					Status:  metav1.ConditionFalse,
-					Reason:  "PodsReady",
+					Reason:  kueue.WorkloadWaitForPodsStart,
```
Done

mimowo commented on test/integration/singlecluster/controller/jobs/job/job_controller_test.go:
```diff
@@ -2068,24 +2069,29 @@ var _ = ginkgo.Describe("Interacting with scheduler", ginkgo.Ordered, ginkgo.Con
 				BackoffBaseSeconds: ptr.To(backoffBaseSeconds),
 				BackoffLimitCount:  backoffLimitCount,
 			},
+			RecoveryTimeout: &metav1.Duration{Duration: util.TinyTimeout},
```
I would suggest a separate config (probably Describe?) for that. Yes, this is more code, but:
- testing without `RecoveryTimeout` makes sense as this remains a valid configuration
- it is hard for reviewers to decouple which changes in the files are due to logic changes and which due to the introduction of the timeout.

mimowo replied:
cc @tenzen-y

mimowo commented on test/integration/singlecluster/controller/jobs/job/job_controller_test.go:
```diff
@@ -2115,34 +2121,122 @@ var _ = ginkgo.Describe("Job controller interacting with Workload controller whe
+	ginkgo.When("long backoffBaseSeconds and tiny waitForPodsReady.timeout", func() {
 		ginkgo.BeforeEach(func() {
 			backoffBaseSeconds = 10
+			waitForPodsReadyTimeout = &metav1.Duration{Duration: util.TinyTimeout}
```
What is the purpose of this change? It is completely non-obvious to me how this is related to the recovery timeout. Please revert if possible. As posted in the other comment, I think it is better to have a bit more of extra code but have a dedicated configuraiton.

PBundyra replied:
Those changes were actually needed, as previously for every `When` scenario the waitForPodsReadyTimeout was hardcoded to tiny, as those test check if the Workload is evicted. Given that I wanted to test eviction because of `.recoveryTimeout` and not `.timeout` I had to prolong it, and this is where this change is coming from

PBundyra replied:
> Plus, if you don't modify configuration for the existing ones you can test what happens if the pod fails, but there is no recovertyTimeoutSet, which actually is quite important to check.

If we want to test this then the changes need to stay

mimowo replied:
Ah, I got it, so you are not changing the configuration. Just setting it to tiny as it was hardcoded before, fine.

mimowo commented on test/integration/singlecluster/controller/jobs/job/job_controller_test.go:
```diff
@@ -2177,40 +2271,40 @@ var _ = ginkgo.Describe("Job controller interacting with Workload controller whe
 		ginkgo.BeforeEach(func() {
 			backoffBaseSeconds = 1
 			backoffLimitCount = ptr.To[int32](1)
+			waitForPodsReadyTimeout = &metav1.Duration{Duration: util.TinyTimeout}
```
Simiarly here, please revert if not needed.

mimowo commented:
> Sorry, it should point to `test/integration/singlecluster/controller/jobs/job/job_controller_test.go` L2182

Yeah, this looks like the scenario I was looking for, but somehow missed it in the amount of changes for integration tests. I posted comments to de-couple the integration tests, and have a new dedicated config. Then, use the old config to confirm that a failed pod does not trigger eviction (as before).

mimowo commented on test/integration/singlecluster/controller/jobs/job/job_controller_test.go:
```diff
@@ -2115,34 +2121,122 @@ var _ = ginkgo.Describe("Job controller interacting with Workload controller whe
 						Status:  metav1.ConditionFalse,
-						Reason:  kueue.WorkloadPodsReady,
-						Message: "Not all pods are ready or succeeded",
+						Reason:  kueue.WorkloadNotAdmitted,
+						Message: "Workload isn't admitted",
+					}, util.IgnoreConditionTimestampsAndObservedGeneration),
+					gomega.BeComparableTo(metav1.Condition{
+						Type:    kueue.WorkloadQuotaReserved,
+						Status:  metav1.ConditionFalse,
+						Reason:  "Pending",
+						Message: fmt.Sprintf("Exceeded the PodsReady timeout %s", wlKey.String()),
+					}, util.IgnoreConditionTimestampsAndObservedGeneration),
+					gomega.BeComparableTo(metav1.Condition{
+						Type:    kueue.WorkloadEvicted,
+						Status:  metav1.ConditionTrue,
+						Reason:  kueue.WorkloadEvictedByPodsReadyTimeout,
+						Message: fmt.Sprintf("Exceeded the PodsReady timeout %s", wlKey.String()),
+					}, util.IgnoreConditionTimestampsAndObservedGeneration),
+					gomega.BeComparableTo(metav1.Condition{
+						Type:    kueue.WorkloadAdmitted,
+						Status:  metav1.ConditionFalse,
+						Reason:  "NoReservation",
+						Message: "The workload has no reservation",
+					}, util.IgnoreConditionTimestampsAndObservedGeneration),
+					gomega.BeComparableTo(metav1.Condition{
+						Type:    kueue.WorkloadRequeued,
+						Status:  metav1.ConditionFalse,
+						Reason:  kueue.WorkloadEvictedByPodsReadyTimeout,
+						Message: fmt.Sprintf("Exceeded the PodsReady timeout %s", wlKey.String()),
+					}, util.IgnoreConditionTimestampsAndObservedGeneration),
```
Please revert to assert only these which were asserted. I think we use ContainElements for the reason - there used to be flakes as some of the other conditions might be transitioning.

mimowo commented on test/integration/singlecluster/controller/jobs/job/job_controller_test.go:
```diff
@@ -2115,34 +2121,122 @@ var _ = ginkgo.Describe("Job controller interacting with Workload controller whe
 			job := testingjob.MakeJob("job", ns.Name).Queue(lq.Name).Request(corev1.ResourceCPU, "2").Obj()
 			gomega.Expect(k8sClient.Create(ctx, job)).Should(gomega.Succeed())
 
-			wl := &kueue.Workload{}
-			wlKey := types.NamespacedName{Name: workloadjob.GetWorkloadNameForJob(job.Name, job.UID), Namespace: job.Namespace}
```
Can you revert? Sorry, I know the cleanup looks ok, but the PR is already large and so the drive-by cleanups add distractions IMO.

PBundyra replied:
Done, sorry for polluting the PR

PBundyra commented:
I believe I addressed all the comments, to sum up the changes since yesterday:
- Removed the `NotAdmitted` reason;
- Added two more integration tests:

Besides unit tests, currently we have 3 integration tests:
1. Checks that if there's no `.recoveryTimeout` a workload doesn't get evicted if a Pod fails (as it was before)
2. Checks that if there's `.recoveryTimeout` and the timeout is surpassed a workload gets evicted if a Pod fails
3. Checks that if there's `.recoveryTimeout` and a workload recovers before the timeout is surpassed, it doesn't get evicted

@mimowo @tenzen-y

mimowo commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -1148,24 +1150,61 @@ func (r *JobReconciler) ignoreUnretryableError(log logr.Logger, err error) error
+
+	case podsReady && podsReadyCond.Reason == kueue.WorkloadWaitForPodsRecovery:
+		return workload.CreatePodsReadyCondition(metav1.ConditionTrue,
+			kueue.WorkloadWaitForPodsRecovery,
```
@PBundyra @tenzen-y I find these reasons `WaitForPodsRecovery` and `WaitForPodsStart` suboptimal for the positive scenarios (when pods are ready). Shouldn't we have dedicated "PodsRecovered" and "PodsStarted" reasons?

PBundyra replied:
I see, added

mimowo commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -1148,24 +1150,61 @@ func (r *JobReconciler) ignoreUnretryableError(log logr.Logger, err error) error
+
+	// a) The workload was running and a pod has failed
+	// b) A pod has failed and the Workload has already recovered once before
+	case !podsReady && ((podsReadyCond.Reason == kueue.WorkloadWaitForPodsStart || podsReadyCond.Reason == "PodsReady") && podsReadyCond.Status == metav1.ConditionTrue ||
+		podsReadyCond.Reason == kueue.WorkloadWaitForPodsRecovery):
```
TBH, I'm still sceptical about the spagetti entry condtitions, and still wasn't able to verify they are complementary, shouldn't it be just simply:

```
case !podsReady && podsReadyCond.Status == metav1.ConditionTrue:
```
meaning we wait for recovery because it was running before and is no longer running?

PBundyra replied:
Simplified

mimowo commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -1148,24 +1150,61 @@ func (r *JobReconciler) ignoreUnretryableError(log logr.Logger, err error) error
+			kueue.WorkloadWaitForPodsStart,
+			"All pods reached readiness and the workload is running")
+
+	case !podsReady && (podsReadyCond.Reason == kueue.WorkloadWaitForPodsStart || podsReadyCond.Reason == "PodsReady") && podsReadyCond.Status == metav1.ConditionFalse:
```
```suggestion
	case !podsReady && podsReadyCond.Status == metav1.ConditionFalse:
```
why not just like this?

PBundyra replied:
Simplified

mimowo commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -1148,24 +1150,60 @@ func (r *JobReconciler) ignoreUnretryableError(log logr.Logger, err error) error
+		"Pods are ready", podsReady)
+
+	if podsReady {
+		var reason string
+		switch {
+		case podsReadyCond == nil:
+			reason = kueue.WorkloadStarted
+
+		case podsReadyCond.Reason == kueue.WorkloadWaitForPodsRecovery:
+			reason = kueue.WorkloadRecovered
+
+		default:
+			reason = kueue.WorkloadStarted
+		}
```
```suggestion
                reason := kueue.WorkloadStarted
		if podsReadyCond != nil && podsReadyCond.Reason == kueue.WorkloadWaitForPodsRecovery {
			reason = kueue.WorkloadRecovered
                }
```
Seems equivalent and much shorter

PBundyra commented on apis/kueue/v1beta1/workload_types.go:
```diff
@@ -587,6 +587,12 @@ const (
 	// when the Pods were ready since the workload admission, but some pod has failed,
 	// and workload waits for recovering.
 	WorkloadWaitForPodsRecovery = "WaitForPodsRecovery"
```
I like this, added

mimowo commented:
LGTM seeing the most recent updates. I will return here when the tests complete.

mimowo commented on test/integration/singlecluster/controller/jobs/job/job_controller_test.go:
```diff
@@ -1028,8 +1029,16 @@ var _ = ginkgo.Describe("When waitForPodsReady enabled", ginkgo.Ordered, ginkgo.
+						}, util.IgnoreConditionTimestampsAndObservedGeneration),
+					))
 				}, util.Timeout, util.Interval).Should(gomega.Succeed())
-				util.SyncAdmittedConditionForWorkloads(ctx, k8sClient, createdWorkload)
```
@PBundyra why is this line removed?

PBundyra replied:
I added the check above which ensures the condition is synced

mimowo replied:
Is it effectively doing the same?

mimowo replied:
Maybe separate is better, not because the change is bad, but because the role of the change is non-obvious in the PR with so many other changes, so it may distract future readers.

PBundyra replied:
I added this because one of the integration test was flaky because the Workload sometimes was still admitted when generating a new PodsReady condition. So this change is directly related to the whole PR

mimowo commented:
/lgtm
/approve
/hold
Leaving final pass to @tenzen-y since there are some changes compared to the KEP

tenzen-y commented:
> /lgtm /approve /hold Leaving final pass to @tenzen-y since there are some changes compared to the KEP

ACK, I will take a look after patch release

tenzen-y commented on test/integration/singlecluster/controller/jobs/job/job_controller_test.go:
I'm ok with the removal of the old `PodsReady` reason tests from other types of Job integration tests like RayJob.
However, It would be worth keeping one test case for the old `PodsReady` reason in this Job integration test since Job integration tests are responsible for verifying JobFramework availability.

tenzen-y replied:
That makes sense. I'm ok without those tests for JobController, thanks.

tenzen-y reviewed (commented):
Basically, lgtm
Additionally, could you open KEP update PR as a follow-up?

PBundyra commented:
> Additionally, could you open KEP update PR as a follow-up?

Absolutely, will do

PBundyra commented on apis/kueue/v1beta1/workload_types.go:
```diff
@@ -579,14 +579,20 @@ const (
+	// WorkloadWaitForStart indicates the reason for PodsReady=False condition
 	// when the pods have not been ready since admission, or the workload is not admitted.
-	WorkloadWaitForPodsStart = "WaitForPodsStart"
+	WorkloadWaitForStart = "WaitForStart"
 
-	// WorkloadWaitForPodsStart indicates the reason for the PodsReady=False condition
+	// WorkloadWaitForRecovery indicates the reason for the PodsReady=False condition
 	// when the Pods were ready since the workload admission, but some pod has failed,
 	// and workload waits for recovering.
-	WorkloadWaitForPodsRecovery = "WaitForPodsRecovery"
+	WorkloadWaitForRecovery = "WaitForRecovery"
+
+	// WorkloadStarted indicates that all Pods are ready and the Workload has successfully started
+	WorkloadStarted = "Started"
+
+	// WorkloadRecovered indicates that after at least one Pod has failed, the Workload has recovered and is running
+	WorkloadRecovered = "Recovered"
```
I happy to improve the documentation in general so this could be another thing to consider. However, please note that the additional two new reasons ("Started"/"Recovered") were introduce to actually ease the debugging progress

tenzen-y replied:
Yeah, here I do not have a strong offering. Just nice to have :)

PBundyra commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -1148,24 +1150,52 @@ func (r *JobReconciler) ignoreUnretryableError(log logr.Logger, err error) error
-		Message: message,
-		// ObservedGeneration is added via workload.UpdateStatus
+func generatePodsReadyCondition(log logr.Logger, job GenericJob, wl *kueue.Workload) metav1.Condition {
+	if !workload.IsAdmitted(wl) {
```
Done

PBundyra commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -460,15 +460,17 @@ func (r *JobReconciler) ReconcileGenericJob(ctx context.Context, req ctrl.Reques
-			log.V(3).Info(fmt.Sprintf("Updating the PodsReady condition with status: %v", condition.Status))
+		condition := generatePodsReadyCondition(log, job, wl)
+		if !workload.HasConditionWithTypeAndReason(wl, &condition) {
+			log.V(3).Info(fmt.Sprintf("Updating the PodsReady condition with reason: %v and status: %v", condition.Reason, condition.Status))
```
Done

PBundyra commented:
> > Additionally, could you open KEP update PR as a follow-up?
> 
> Absolutely, will do

Follow up with KEP update:
https://github.com/kubernetes-sigs/kueue/pull/4419

PBundyra commented:
/hold cancel
As I got two approvals
